src/engine/renderer/tr_image.cpp

```diff
@@ -2790,15 +2794,79 @@ void R_CreateBuiltinImages()
 	R_CreateColorGradeImage();
 }
 
+void R_CreateCharsetImage()
+{
+	// charsetImage
+	if ( TTF_Init() < 0 )
+	{
+		Log::Warn( "R_CreateCharsetImage: could not load init SDL_ttf");
+		tr.charsetImage = tr.whiteImage;
+		return;
+	}
+	SDL_Surface* charsetSurf = SDL_CreateRGBSurface ( SDL_SWSURFACE, 16*CHARSETIMAGE_CHARSIZE, 16*CHARSETIMAGE_CHARSIZE, 32, 0xFF000000, 0x00FF0000, 0x0000FF00, 0x000000FF );
+	char* fontname = Cvar_Get( "cl_consoleFont", "fonts/unifont.ttf",  CVAR_LATCH )->string;
+	std::string fontfile = FS::PakPath::ReadFile ( fontname );
+	TTF_Font* font = TTF_OpenFontRW ( SDL_RWFromConstMem ( fontfile.c_str (), fontfile.size () ), 1, CHARSETIMAGE_CHARSIZE );
+	if ( font == NULL )
+	{
+		Log::Warn( "R_CreateCharsetImage: could not open font %s", fontname );
+		tr.charsetImage = tr.whiteImage;
+		return;
+	}
+	SDL_Surface* charSurf = NULL;
+	SDL_Rect dest;
+	dest.w = 16;
+	dest.h = 16;
+	char c[2];
+	c[1] = 0;
+	for (uint8_t i = 0; i < 16; ++i)
+	{
+		for (uint8_t j = 0; j < 16; ++j)
+		{
+			dest.x = j*CHARSETIMAGE_CHARSIZE;
+			dest.y = i*CHARSETIMAGE_CHARSIZE;
+			c[0] = i*16+j;
+			charSurf = TTF_RenderText_Solid ( font, c, {255, 255, 255} );
+			SDL_BlitSurface ( charSurf, NULL, charsetSurf, &dest );
+			SDL_FreeSurface ( charSurf );
+		}
+	}
+	SDL_LockSurface ( charsetSurf );
+	byte  data[ CHARSETIMAGE_CHARSIZE*16 ][ CHARSETIMAGE_CHARSIZE*16 ][ 4 ];
+	byte  *dataPtr = &data[0][0][0];
+	byte  *out = &data[ 0 ][ 0 ][ 0 ];
```

Comment:
Any reason data is declared as a 3D array but used as a flat array?
You could just declare byte  `data[ CHARSETIMAGE_CHARSIZE*16 * CHARSETIMAGE_CHARSIZE*16 * 4 ];`